pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -449,156 +490,185 @@ func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hi
-			}
+	return reconcile.Result{}, nil
+}
 
-			jobDeleted, err := r.deleteJobOnHashChange(existingJob, job, cdLog)
-			if jobDeleted || err != nil {
-				return reconcile.Result{}, err
-			}
-		}
+func (r *ReconcileClusterDeployment) reconcileExistingProvision(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (result reconcile.Result, returnedErr error) {
+	cdLog = cdLog.WithField("provision", cd.Status.Provision.Name)
+	cdLog.Debug("reconciling existing provision")
 
-		if firstInstalledObserve && cd.Status.InstalledTimestamp == nil {
-			now := metav1.Now()
-			cd.Status.InstalledTimestamp = &now
-		}
+	provision := &hivev1.ClusterProvision{}
+	switch err := r.Get(context.TODO(), types.NamespacedName{Name: cd.Status.Provision.Name, Namespace: cd.Namespace}, provision); {
+	case apierrors.IsNotFound(err):
+		cdLog.Warn("linked provision not found")
+		return r.clearOutCurrentProvision(cd, cdLog)
+	case err != nil:
+		cdLog.WithError(err).Error("could not get provision")
+		return reconcile.Result{}, err
 	}
 
-	err = r.updateClusterDeploymentStatus(cd, origCD, existingJob, cdLog)
-	if err != nil {
-		cdLog.WithError(err).Errorf("error updating cluster deployment status")
+	// Save the cluster ID and infra ID from the provision so that we can
+	// clean up partial installs on the next provision attempt in case of failure.
+	if (provision.Spec.ClusterID != nil && cd.Status.ClusterID != *provision.Spec.ClusterID) ||
+		(provision.Spec.InfraID != nil && cd.Status.InfraID != *provision.Spec.InfraID) {
+		cdLog.Infof("Saving infra ID %q for cluster", *provision.Spec.InfraID)
+		if provision.Spec.ClusterID != nil {
+			cd.Status.ClusterID = *provision.Spec.ClusterID
+		}
+		if provision.Spec.InfraID != nil {
+			cd.Status.InfraID = *provision.Spec.InfraID
+		}
+		err := r.Status().Update(context.TODO(), cd)
 		return reconcile.Result{}, err
 	}
 
-	// firstInstalledObserve will be true if this is the first time we've noticed the install job completed.
-	// If true, we know we can report the metrics associated with a completed job.
-	if firstInstalledObserve {
-		// jobDuration calculates the time elapsed since the install job started
-		jobDuration := existingJob.Status.CompletionTime.Time.Sub(existingJob.Status.StartTime.Time)
+	switch provision.Spec.Stage {
+	case hivev1.ClusterProvisionStageProvisioning:
+		cdLog.Debug("still provisioning")
+		return reconcile.Result{}, nil
+	case hivev1.ClusterProvisionStageFailed:
+		nextProvisionTime := calculateNextProvisionTime(provision, cdLog)
+		timeUntilNextProvision := time.Until(nextProvisionTime)
+		if timeUntilNextProvision.Seconds() > 0 {
+			cdLog.WithField("next provision", nextProvisionTime).Info("waiting to start a new provision after failure")
+			return reconcile.Result{RequeueAfter: timeUntilNextProvision}, nil
+		}
+		cdLog.Info("clearing current failed provision to make way for a new provision")
+		return r.clearOutCurrentProvision(cd, cdLog)
+	case hivev1.ClusterProvisionStageComplete:
```

Comment:
Absolutely not for this PR, but I kinda feel like there's a lot of ClusterProvision processing code in the ClusterDeployment controller. Do you think some of this would be reasonable to refactor out to the ClusterProvision controller in future? Not sure if you feel this is the best place for it, or if it's just how it evolved given the job code it replaced.

Reply:
Hmm, I'm not sure that I feel the same way as you. The code around the area of the comment is doing things on the ClusterDeployment, which makes it appropriate for the ClusterDeployment controller. The ClusterProvision controller reconciles ClusterProvisions and should not care about the ClusterDeployment. In fact, you may notice that the ClusterProvision controller does not even get the ClusterDeployment associated with the ClusterProvision.